orchagent/fdborch.cpp

```diff
@@ -63,3 +70,218 @@ bool FdbOrch::getPort(const MacAddress& mac, uint16_t vlan, Port& port)
+    attrs.push_back(attr);
+
+    attr.id = SAI_FDB_ENTRY_ATTR_PORT_ID;
+    attr.value.oid = port.m_port_id;
+    attrs.push_back(attr);
+
+    attr.id = SAI_FDB_ENTRY_ATTR_PACKET_ACTION;
+    attr.value.u8 = SAI_PACKET_ACTION_FORWARD; // FIXME: what to use here? what about u8 type?
+    attrs.push_back(attr);
+
+    status = sai_fdb_api->create_fdb_entry(&fdb_entry, attrs.size(), attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to add FDB entry. mac=%s, vlan=%d. port_name %s. type %s\n",
+                   entry.mac.to_string().c_str(), entry.vlan, port_name.c_str(), type.c_str());
+        return false; // FIXME: retry it? How many times?
+    }
+
+    (void)m_entries.insert(entry);
+
+    return true;
+}
+
+bool FdbOrch::removeFdbEntry(const FdbEntry& entry)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_DEBUG("mac=%s, vlan=%d\n", entry.mac.to_string().c_str(), entry.vlan);
+
+    if (m_entries.count(entry) == 0)
+    {
+        SWSS_LOG_ERROR("FDB entry isn't found. mac=%s vlan=%d\n", entry.mac.to_string().c_str(), entry.vlan);
+        return true;
+    }
+
+    sai_status_t status;
+    sai_fdb_entry_t fdb_entry;
+    memcpy(fdb_entry.mac_address, entry.mac.getMac(), sizeof(sai_mac_t));
+    fdb_entry.vlan_id = entry.vlan;
+
+    status = sai_fdb_api->remove_fdb_entry(&fdb_entry);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to remove FDB entry. mac=%s, vlan=%d\n",
+                       entry.mac.to_string().c_str(), entry.vlan);
+        return false; // FIXME: retry it? How many times?
+    }
+
+    (void)m_entries.erase(entry);
+
+    return true;
+}
+
+bool FdbOrch::splitKey(const string& key, FdbEntry& entry)
+{
+    SWSS_LOG_ENTER();
+
+    string mac_address_str;
+    string vlan_str;
+
+    size_t found = key.rfind(':');
+    if (found == string::npos)
+    {
+        SWSS_LOG_ERROR("Failed to parse key: %s\n", key.c_str());
+        return false;
+    }
+
+    mac_address_str = key.substr(0, found);
+    vlan_str = key.substr(found + 1, string::npos); // FIXME: what if we have "macaddress:"
+    if (vlan_str.length() <= 4) // "Vlan"
+    {
+        SWSS_LOG_ERROR("Failed to extract vlan interface name from the key: %s\n", key.c_str());
+        return false;
+    }
+
+    uint8_t mac_array[6];
+    if (!MacAddress::parseMacString(mac_address_str, mac_array)) // FIXME: parseMacString consider 52:54:00:25::E9 and 52:54:00:25:E9 as good one
+    {
+        SWSS_LOG_ERROR("Failed to parse mac address: %s in key: %s\n", mac_address_str.c_str(), key.c_str());
+        return false;
+    }
+
+    entry.mac = MacAddress(mac_array);
+
+    Port port;
+    if (!m_portsOrch->getPort(vlan_str, port))
+    {
+        SWSS_LOG_ERROR("Failed to get port for %s\n", vlan_str.c_str());
+        return false;
+    }
+
+    entry.vlan = stoi(vlan_str.substr(4)); // FIXME: create swss-common function to extract vlan number from vlan name
```

Comment:
It could be added to swssnet.h file in sonic-swss, as it will not be used outside swss.